src/Microsoft.AspNetCore.Http.Abstractions/Routing/RouteValueDictionary.cs

```diff
@@ -456,7 +488,10 @@ private void EnsureCapacitySlow(int capacity)
             if (_propertyStorage != null)
             {
                 var storage = _propertyStorage;
-                capacity = Math.Max(storage.Properties.Length, capacity);
+                
+                // If we're converting from properties, it's likely due to an 'add' to make sure we have at least
+                // the default amount of space.
+                capacity = Math.Max(DefaultCapacity, Math.Max(storage.Properties.Length, capacity));
```

Comment:
This was a bug I discovered while testing TryAdd. If we 'upgrade' from property storage to array we used to allocate the minimum number of slots. Since upgrade is usually paired with adding an item, the common case is that you would allocate and then immediately resize.